File: examples/testlib/art/cartoon/stickman/hand.js
const defaultStyle = { stroke: "#f97316", width: 0.8 };
const { toPoint: toPointOrNull, normalizeInput } = helpers;

function extractJointPoints(config) {
  const joints = config.joints;
  if (!joints || typeof joints !== "object") {
    return null;
  }
  const attachment = toPointOrNull(joints.attachment ?? joints.shoulder);
  const hinge = toPointOrNull(joints.hinge ?? joints.elbow);
  const effector = toPointOrNull(joints.effector ?? joints.wrist);
  if (attachment && hinge && effector) {
    return {
      attachment,
      hinge,
      effector
    };
  }
  return null;
}

function renderFromJoints(style, joints) {
  const segments = [
    {
      type: "line",
      from: joints.attachment,
      to: joints.hinge,
      stroke: style.stroke,
      width: style.width
    },
    {
      type: "line",
      from: joints.hinge,
      to: joints.effector,
      stroke: style.stroke,
      width: style.width
    }
  ];

  return {
    graphics: segments
  };
}

function createHand(options) {
  const config = normalizeInput(options, {});
  const joints = extractJointPoints(config);
  if (!joints) {
    throw new Error(
      "hand model requires helpers.toPoint and skeleton-provided joints (hand.joints missing or helpers.toPoint undefined)"
    );
  }

  const style = {
    ...defaultStyle,
    ...(config.style && typeof config.style === "object" ? config.style : {})
  };

  return renderFromJoints(style, joints);
}

return createHand;

File: examples/testlib/art/cartoon/stickman/hand.fs
(options)=>
{
  defaults:{ stroke:"#f97316"; width:0.8 };

  config:options ?? {};
  joints:extractJointPoints(config);

  eval if joints = null then {
    graphics:[];
  } else {
    style:config.style ?? {};
    stroke:style.stroke ?? defaults.stroke;
    width:style.width ?? defaults.width;

    lines:[
      {
        type:"line";
        from:joints.attachment;
        to:joints.hinge;
        stroke:stroke;
        width:width;
      },
      {
        type:"line";
        from:joints.hinge;
        to:joints.effector;
        stroke:stroke;
        width:width;
      }
    ];

    eval {
      graphics:[{
        type:"testlib/cartoon/stickman/hand";
        name:"testlib/cartoon/stickman/hand";
        graphics:lines;
      }];
    };
  };

  extractJointPoints:(config)=> {
    jointsInput:config.joints;
    eval if jointsInput = null then null else {
      attachment:jointsInput.attachment ?? jointsInput.shoulder;
      hinge:jointsInput.hinge ?? jointsInput.elbow;
      effector:jointsInput.effector ?? jointsInput.wrist;
      eval if attachment = null or hinge = null or effector = null then null else {
        attachment:attachment;
        hinge:hinge;
        effector:effector;
      };
    };
  };
}
